src/app/core/auth/state/auth-state.service.ts

```diff
@@ -0,0 +1,76 @@
+import { Action, State, StateContext } from '@ngxs/store'
+import { Injectable } from '@angular/core'
+import { AuthService } from 'src/app/core/auth/services/auth.service'
+import { Observable, tap } from 'rxjs'
+import { Tokens } from 'src/app/core/auth/types/tokens'
+import { TokensStorageService } from "../services/tokens-storage.service";
+import { Login, LoginOut, SetTokens, UpdateLoginFormSubmission } from "./auth.actions";
+import { Router } from "@angular/router";
+
+export interface AuthStateModel {
+    tokens?: Tokens,
+    isAuthenticated?: boolean
+    isLoginFormSubmitted?: boolean
+}
+
+@State<AuthStateModel>({
+    name: 'auth',
+    defaults: {
+        tokens: null,
+        isAuthenticated: false,
+        isLoginFormSubmitted: false
+    }
+})
+
+@Injectable()
+export class AuthStateService {
+    constructor(
+        private readonly _authService: AuthService,
+        private readonly _tokenStorageService: TokensStorageService,
+        private readonly _router: Router
+    ) {
+    }
+
+    @Action(Login)
+    public login(
+        { patchState, dispatch }: StateContext<AuthStateModel>,
+        { credentials, remember }: Login
+    ): Observable<Tokens> {
+        return this._authService.getToken(credentials).pipe(
+            tap((tokens: Tokens) => {
```

Comment:
You do not need to specify types like this in most cases. Typescript [infers](https://www.typescriptlang.org/docs/handbook/type-inference.html) types very well and already knows what are the types here.